File: TuCSoN/trunk/src/alice/tucson/examples/spawnedWorkers/SpawnedWorkingActivity.java
package alice.tucson.examples.spawnedWorkers;

import java.math.BigInteger;
import alice.logictuple.LogicTuple;
import alice.logictuple.TupleArgument;
import alice.logictuple.exceptions.InvalidLogicTupleException;
import alice.tucson.api.AbstractSpawnActivity;
import alice.tucson.api.exceptions.TucsonInvalidLogicTupleException;
import alice.tuplecentre.api.exceptions.InvalidOperationException;

/**
 *
 *
 * @author ste (mailto: s.mariani@unibo.it)
 */
public class SpawnedWorkingActivity extends AbstractSpawnActivity {

    private static final long serialVersionUID = -4459068799410719933L;

    /*
     *
     */
    @Override
    public void doActivity() {
        try {
            /*
             * Jobs collection phase.
             */
            final LogicTuple templ = LogicTuple
                    .parse("fact(master(M),num(N),reqID(R))");
            this.log("Waiting for jobs...");
            /*
             * Watch out: it's a suspensive primitive! If no jobs are available
             * we are stuck!
             */
            final LogicTuple job = this.in(templ);
            this.log("Found job: " + job.toString());
            /*
             * Computation phase.
             */
            final BigInteger bigNum = this.computeFactorial(job.getArg("num")
                    .getArg(0));
            /*
             * Result submission phase.
             */
            final LogicTuple res = LogicTuple.parse("res(" + "master("
                    + job.getArg("master").getArg(0) + ")," + "fact("
                    + bigNum.toString() + ")," + "reqID("
                    + job.getArg("reqID").getArg(0) + ")" + ")");
            this.log("Putting result: " + res.toString());
            this.out(res);
        } catch (final InvalidLogicTupleException e) {
            this.log("ERROR: Tuple is not an admissible Prolog term!");
            e.printStackTrace();
        } catch (final InvalidOperationException e) {
            this.log("ERROR: No tuple arguments to retrieve!");
            e.printStackTrace();
        }
    }

    private BigInteger computeFactorial(final TupleArgument varValue) {
        try {
            final int num = varValue.intValue();
            this.log("Computing factorial for: " + num + "...");
            return this.factorial(num);
        } catch (final InvalidOperationException e) {
            this.log("Not an Integer value, killing myself...");
            return new BigInteger("-1");
        }
    }

    private BigInteger factorial(final int num) {
        if (num == 0) {
            return BigInteger.ONE;
        }
        return new BigInteger(String.valueOf(num)).multiply(this
                .factorial(num - 1));
    }
}

File: TuCSoN/classes/alice/tucson/examples/spawnedWorkers/spawnedWorkingActivity.pl
% doActivity(+Tc):- performs the factorial computation reading jobs from
% tuple centre <Tc>.
doActivity(Tc):-
    write("Waiting for jobs in <"), write(Tc), write(">..."), nl,
    in(fact(num(Fact))),
    write("Computing factorial for <"), write(Fact), write(">..."), nl,
    computeFact(Fact, Res),
    write("Computed factorial <"), write(Res), write(">."), nl,
    out(res(fact(Res))).

% computeFact(+Fact, -Res):- computes factorial of <Fact>, storing the
% result in <Res>.
computeFact(Fact, Res):-
    computeFact(Fact, Res, 1).

    computeFact(0, Res, Res):- !.
    computeFact(1, Res, Res):- !.
    computeFact(Fact, Res, Temp):-
        Fact2 is Fact-1, Temp2 is Fact * Temp,
        computeFact(Fact2, Res, Temp2).
